File: src/kairos/detectors/resonance.py
"""Multi-dimension resonance scorer with Signal Quality Score.

Signal Quality Score (0-100):
  - Extremity (0-40): max Z-score across dimensions
  - Resonance (0-30): 2→10, 3→20, 4+→30
  - Direction (0-20): all dimensions point same way
  - Context (0-10): funding + OI combo

  Push if score ≥ min_score (default 55). Tune per config.
"""

from __future__ import annotations

import logging
import time
import threading
from dataclasses import dataclass, field
from typing import Any, Callable

from kairos.detectors.base import AnomalyEvent

logger = logging.getLogger(__name__)


@dataclass
class ResonanceEvent:
    """A multi-dimension resonance-scored alert."""

    symbol: str
    signal_score: float
    dimension_count: int
    dimensions: dict[str, AnomalyEvent]
    timestamp: float = field(default_factory=time.time)

    def to_alert_event(self) -> AnomalyEvent:
        data: dict[str, Any] = {
            "signal_score": self.signal_score,
            "dimension_count": self.dimension_count,
            "dimensions": list(self.dimensions.keys()),
        }
        for dim, ev in self.dimensions.items():
            data[f"{dim}_data"] = ev.data
        return AnomalyEvent(
            symbol=self.symbol,
            event_type="resonance",
            severity="HIGH",
            data=data,
            timestamp=self.timestamp,
        )


class ResonanceScorer:
    """Aggregates AnomalyEvents from detectors. Pushes when score ≥ min_score."""

    def __init__(self, config: dict | None = None) -> None:
        cfg = config or {}
        rs = cfg.get("resonanceScorer", cfg)
        self.enabled = bool(rs.get("enabled", True))
        self.window_seconds = float(rs.get("windowSeconds", 300))
        self.min_dimensions = int(rs.get("minDimensions", 2))
        self.min_score = float(rs.get("minScore", 55))
        self.cooldown_seconds = float(rs.get("cooldownSeconds", 600))

        self._windows: dict[str, dict[str, AnomalyEvent]] = {}
        self._last_emission: dict[str, float] = {}
        self._callbacks: list[Callable[[ResonanceEvent], None]] = []
        self._lock = threading.RLock()

    def on_event(self, event: AnomalyEvent) -> None:
        if not self.enabled or event.event_type == "resonance":
            return
        with self._lock:
            now = event.timestamp or time.time()
            self._prune_windows(now)
            symbol = event.symbol
            if symbol not in self._windows:
                self._windows[symbol] = {}
            existing = self._windows[symbol].get(event.event_type)
            if existing and _extremity_z(existing.data) >= _extremity_z(event.data):
                return
            self._windows[symbol][event.event_type] = event
            self._evaluate(symbol, now)

    def set_callback(self, callback: Callable[[ResonanceEvent], None]) -> None:
        self._callbacks.append(callback)

    def _prune_windows(self, now: float) -> None:
        cutoff = now - self.window_seconds
        expired = [sym for sym, dims in self._windows.items()
                   if all(e.timestamp < cutoff for e in dims.values())]
        for sym in expired:
            del self._windows[sym]

    def _evaluate(self, symbol: str, now: float) -> None:
        dims = self._windows.get(symbol)
        if not dims or len(dims) < self.min_dimensions:
            return
        if now - self._last_emission.get(symbol, 0.0) < self.cooldown_seconds:
            return

        score = _signal_quality_score(dims)
        if score < self.min_score:
            return

        r = ResonanceEvent(symbol=symbol, signal_score=round(score),
                           dimension_count=len(dims), dimensions=dict(dims),
                           timestamp=now)
        self._last_emission[symbol] = now
        for cb in self._callbacks:
            try:
                cb(r)
            except Exception as exc:
                logger.error("ResonanceScorer callback error: %s", exc)

    def reset(self) -> None:
        with self._lock:
            self._windows.clear()
            self._last_emission.clear()


# ── Signal Quality Score ──────────────────────────────────────────

def _signal_quality_score(dims: dict[str, AnomalyEvent]) -> float:
    max_z = max((_extremity_z(e.data) for e in dims.values()), default=0.0)
    if max_z >= 5.0:
        extremity = 40
    elif max_z >= 4.0:
        extremity = 30
    elif max_z >= 3.0:
        extremity = 20
    elif max_z >= 2.5:
        extremity = 10
    else:
        max_sev = max((_sev_score(e.severity) for e in dims.values()), default=1)
        extremity = (max_sev - 1) * 10

    n = len(dims)
    resonance = min(n, 4) * 10 - 10

    directions = [_direction_bias(e.event_type, e.data) for e in dims.values()]
    non_neutral = [d for d in directions if d != 0]
    if len(non_neutral) >= 2 and all(d > 0 for d in non_neutral):
        direction_score = 20
    elif len(non_neutral) >= 2 and all(d < 0 for d in non_neutral):
        direction_score = 20
    elif len(non_neutral) >= 2:
        has_funding = "funding_rate_anomaly" in dims
        has_ls = "long_short_ratio" in dims
        has_liq = "liquidation" in dims
        if has_funding and has_ls:
            f_dir = _direction_bias("funding_rate_anomaly", dims["funding_rate_anomaly"].data)
            ls_dir = _direction_bias("long_short_ratio", dims["long_short_ratio"].data)
            direction_score = 10 if (f_dir != 0 and ls_dir != 0 and f_dir == ls_dir) else 0
        elif has_liq:
            direction_score = 10
        else:
            direction_score = 0
    else:
        direction_score = 10

    ctx_bonus = 0
    if "funding_rate_anomaly" in dims and "open_interest_change" in dims:
        ctx_bonus = 5  # funding + OI = crowded trade confirmation

    return min(extremity + resonance + direction_score + ctx_bonus, 100.0)


def _extremity_z(data: dict) -> float:
    z = _get_float(data, "zscore")
    if z is not None and z > 0:
        return abs(z)
    rate = abs(_get_float_or_zero(data, "funding_rate"))
    if rate >= 0.002:
        return 6.0
    if rate >= 0.001:
        return 4.0
    if rate >= 0.0005:
        return 2.5
    millions = _get_float_or_zero(data, "total_liquidation_millions")
    if millions >= 20:
        return 6.0
    if millions >= 10:
        return 4.0
    if millions >= 5:
        return 3.0
    oi = abs(_get_float_or_zero(data, "change_pct"))
    if oi >= 30:
        return 5.0
    if oi >= 15:
        return 3.0
    vel = abs(_get_float_or_zero(data, "change_pct"))
    if vel >= 3:
        return 4.0
    if vel >= 1.5:
        return 2.5
    vol = _get_float_or_zero(data, "ratio")
    if vol >= 15:
        return 5.0
    if vol >= 10:
        return 3.0
    ratio = _get_float(data, "ls_ratio")
    if ratio and abs(ratio - 1) > 3:
        return 5.0
    if data.get("long_rate", 0) >= 85:
        return 4.0
    if data.get("long_rate", 0) >= 80:
        return 2.5
    return 0.0


def _direction_bias(event_type: str, data: dict) -> int:
    if event_type == "price_velocity":
        pct = _get_float_or_zero(data, "change_pct")
        return 1 if pct > 0 else -1
    if event_type == "volume_spike":
        return 0
    if event_type == "open_interest_change":
        return 0
    if event_type == "funding_rate_anomaly":
        rate = _get_float_or_zero(data, "funding_rate")
        return -1 if rate > 0 else 1
    if event_type == "long_short_ratio":
        long_r = _get_float_or_zero(data, "long_rate")
        return -1 if long_r > 60 else (1 if long_r < 40 else 0)
    if event_type == "liquidation":
        long_pct = _get_float_or_zero(data, "long_liquidation_pct")
        return -1 if long_pct > 60 else (1 if long_pct < 40 else 0)
    return 0


def _get_float(data: dict, key: str, default: float | None = None) -> float | None:
    v = data.get(key)
    if v is None:
        return default
    try:
        return float(v)
    except (TypeError, ValueError):
        return default


def _get_float_or_zero(data: dict, key: str) -> float:
    v = data.get(key)
    if v is None:
        return 0.0
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0.0


def _sev_score(s: str) -> int:
    return {"LOW": 1, "MEDIUM": 2, "HIGH": 3}.get(s.strip().upper(), 1)

File: internal/detector/resonance.go
package detector

import (
	"log/slog"
	"math"
	"sync"
	"time"

	"github.com/ArchdevilForge/kairos/internal/types"
)

// ResonanceEvent is a multi-dimension resonance-scored alert.
type ResonanceEvent struct {
	Symbol         string                       `json:"symbol" yaml:"symbol"`
	SignalScore    float64                      `json:"signal_score" yaml:"signal_score"`
	DimensionCount int                          `json:"dimension_count" yaml:"dimension_count"`
	Dimensions     map[string]types.AnomalyEvent `json:"dimensions" yaml:"dimensions"`
	Timestamp      float64                      `json:"timestamp" yaml:"timestamp"`
}

// ToAlertEvent converts the ResonanceEvent to a standard AnomalyEvent.
func (r *ResonanceEvent) ToAlertEvent() types.AnomalyEvent {
	dimKeys := make([]string, 0, len(r.Dimensions))
	for k := range r.Dimensions {
		dimKeys = append(dimKeys, k)
	}
	data := map[string]any{
		"signal_score":    r.SignalScore,
		"dimension_count": r.DimensionCount,
		"dimensions":      dimKeys,
	}
	for dim, ev := range r.Dimensions {
		data[dim+"_data"] = ev.Data
	}
	return types.AnomalyEvent{
		Symbol:    r.Symbol,
		EventType: "resonance",
		Severity:  types.SeverityHigh,
		Data:      data,
		Timestamp: r.Timestamp,
	}
}

// ResonanceScorer aggregates AnomalyEvents from multiple detectors.
// When multiple dimensions fire for the same symbol within a window,
// it computes a Signal Quality Score (0-100) and emits a ResonanceEvent.
//
// Score components (port from Python _signal_quality_score):
//   - Extremity (0-40): max z-score across active dimensions
//   - Resonance (0-30): 2 dimensions → 10, 3 → 20, 4+ → 30
//   - Direction (0-20): all point same way
//   - Context (0-10): funding + OI combo
//
// Push if score ≥ minScore (default 55).
// Window: configurable (default 300s)
// Cooldown: configurable (default 600s)
// Minimum dimensions: configurable (default 2)
// Minimum score: configurable (default 55)
type ResonanceScorer struct {
	enabled         bool
	windowSeconds   float64
	minDimensions   int
	minScore        float64
	cooldownSeconds float64

	mu           sync.RWMutex
	windows      map[string]map[string]types.AnomalyEvent // symbol -> event_type -> event
	lastEmission map[string]float64                       // symbol -> timestamp
	callbacks    []func(ResonanceEvent)
	logger       *slog.Logger
	events       chan ResonanceEvent
}

// NewResonanceScorer creates a ResonanceScorer from config.
func NewResonanceScorer(cfg types.ResonanceScorerConfig) *ResonanceScorer {
	rs := &ResonanceScorer{
		enabled:         cfg.Enabled,
		windowSeconds:   300,
		minDimensions:   2,
		minScore:        55,
		cooldownSeconds: 600,
		windows:         make(map[string]map[string]types.AnomalyEvent),
		lastEmission:    make(map[string]float64),
		logger:          slog.Default().With("detector", "resonance"),
		events:          make(chan ResonanceEvent, 64),
	}
	if cfg.WindowSeconds > 0 {
		rs.windowSeconds = float64(cfg.WindowSeconds)
	}
	if cfg.MinDimensions > 0 {
		rs.minDimensions = cfg.MinDimensions
	}
	if cfg.MinScore > 0 {
		rs.minScore = cfg.MinScore
	}
	if cfg.CooldownSeconds > 0 {
		rs.cooldownSeconds = float64(cfg.CooldownSeconds)
	}
	return rs
}

// OnEvent processes an incoming AnomalyEvent from any detector.
func (rs *ResonanceScorer) OnEvent(event types.AnomalyEvent) {
	if !rs.enabled || event.EventType == "resonance" {
		return
	}
	rs.mu.Lock()
	defer rs.mu.Unlock()

	now := event.Timestamp
	if now <= 0 {
		now = float64(time.Now().UnixMilli()) / 1000
	}

	rs.pruneWindows(now)

	symbol := event.Symbol
	if rs.windows[symbol] == nil {
		rs.windows[symbol] = make(map[string]types.AnomalyEvent)
	}
	existing, ok := rs.windows[symbol][event.EventType]
	if ok && extremityZ(existing.Data) >= extremityZ(event.Data) {
		return
	}
	rs.windows[symbol][event.EventType] = event
	rs.evaluate(symbol, now)
}

// SetCallback registers a callback for ResonanceEvents.
func (rs *ResonanceScorer) SetCallback(cb func(ResonanceEvent)) {
	rs.callbacks = append(rs.callbacks, cb)
}

// Events returns the read-only resonance event channel.
func (rs *ResonanceScorer) Events() <-chan ResonanceEvent {
	return rs.events
}

// pruneWindows removes symbols where all events are older than the window.
func (rs *ResonanceScorer) pruneWindows(now float64) {
	cutoff := now - rs.windowSeconds
	for sym, dims := range rs.windows {
		allExpired := true
		for _, e := range dims {
			if e.Timestamp >= cutoff {
				allExpired = false
				break
			}
		}
		if allExpired {
			delete(rs.windows, sym)
		}
	}
}

// evaluate checks whether the symbol's window meets criteria and emits if so.
func (rs *ResonanceScorer) evaluate(symbol string, now float64) {
	dims := rs.windows[symbol]
	if len(dims) < rs.minDimensions {
		return
	}
	if now-rs.lastEmission[symbol] < rs.cooldownSeconds {
		return
	}

	score := signalQualityScore(dims)
	if score < rs.minScore {
		return
	}

	dimCopy := make(map[string]types.AnomalyEvent, len(dims))
	for k, v := range dims {
		dimCopy[k] = v
	}
	rEvent := ResonanceEvent{
		Symbol:         symbol,
		SignalScore:    math.Round(score),
		DimensionCount: len(dims),
		Dimensions:     dimCopy,
		Timestamp:      now,
	}
	rs.lastEmission[symbol] = now

	// Channel delivery (non-blocking).
	select {
	case rs.events <- rEvent:
	default:
		rs.logger.Warn("resonance event channel full, dropping event")
	}

	// Callback delivery.
	for _, cb := range rs.callbacks {
		func(cb func(ResonanceEvent), ev ResonanceEvent) {
			defer func() {
				if r := recover(); r != nil {
					rs.logger.Error("resonance callback panic", "recover", r)
				}
			}()
			cb(ev)
		}(cb, rEvent)
	}
}

// Reset clears all windows and emission history.
func (rs *ResonanceScorer) Reset() {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	clear(rs.windows)
	clear(rs.lastEmission)
}

// UpdateConfig hot-reloads configuration at runtime.
func (rs *ResonanceScorer) UpdateConfig(cfg types.ResonanceScorerConfig) {
	rs.mu.Lock()
	defer rs.mu.Unlock()
	rs.enabled = cfg.Enabled
	if cfg.WindowSeconds > 0 {
		rs.windowSeconds = float64(cfg.WindowSeconds)
	}
	if cfg.MinDimensions > 0 {
		rs.minDimensions = cfg.MinDimensions
	}
	if cfg.MinScore > 0 {
		rs.minScore = cfg.MinScore
	}
	if cfg.CooldownSeconds > 0 {
		rs.cooldownSeconds = float64(cfg.CooldownSeconds)
	}
}

// ── Signal Quality Score ────────────────────────────────────────────────

// signalQualityScore computes the composite Signal Quality Score (0-100).
func signalQualityScore(dims map[string]types.AnomalyEvent) float64 {
	// 1. Extremity (0-40): max z-score equivalent across dimensions.
	maxZ := 0.0
	for _, e := range dims {
		z := extremityZ(e.Data)
		if z > maxZ {
			maxZ = z
		}
	}

	var extremity float64
	switch {
	case maxZ >= 5.0:
		extremity = 40
	case maxZ >= 4.0:
		extremity = 30
	case maxZ >= 3.0:
		extremity = 20
	case maxZ >= 2.5:
		extremity = 10
	default:
		maxSev := 1
		for _, e := range dims {
			if s := sevScore(string(e.Severity)); s > maxSev {
				maxSev = s
			}
		}
		extremity = float64(maxSev-1) * 10
	}

	// 2. Resonance (0-30): 2→10, 3→20, 4+→30.
	n := len(dims)
	resonance := float64(min(n, 4)*10 - 10)

	// 3. Direction (0-20): consensus among non-neutral dimensions.
	directions := make([]int, 0, n)
	for _, e := range dims {
		directions = append(directions, directionBias(e.EventType, e.Data))
	}
	nonNeutral := make([]int, 0, len(directions))
	for _, d := range directions {
		if d != 0 {
			nonNeutral = append(nonNeutral, d)
		}
	}

	var directionScore float64
	if len(nonNeutral) >= 2 && allPositive(nonNeutral) {
		directionScore = 20
	} else if len(nonNeutral) >= 2 && allNegative(nonNeutral) {
		directionScore = 20
	} else if len(nonNeutral) >= 2 {
		_, hasFunding := dims["funding_rate_anomaly"]
		_, hasLS := dims["long_short_ratio"]
		_, hasLiq := dims["liquidation"]
		if hasFunding && hasLS {
			fDir := directionBias("funding_rate_anomaly", dims["funding_rate_anomaly"].Data)
			lsDir := directionBias("long_short_ratio", dims["long_short_ratio"].Data)
			if fDir != 0 && lsDir != 0 && fDir == lsDir {
				directionScore = 10
			} else {
				directionScore = 0
			}
		} else if hasLiq {
			directionScore = 10
		} else {
			directionScore = 0
		}
	} else {
		directionScore = 10
	}

	// 4. Context bonus (0-10): funding + OI = crowded trade confirmation.
	ctxBonus := 0.0
	_, hasFunding := dims["funding_rate_anomaly"]
	_, hasOI := dims["open_interest_change"]
	if hasFunding && hasOI {
		ctxBonus = 5
	}

	total := extremity + resonance + directionScore + ctxBonus
	if total > 100 {
		total = 100
	}
	return total
}

// extremityZ maps data fields to z-score equivalents.
func extremityZ(data map[string]any) float64 {
	// Direct zscore field.
	if z, ok := getFloat(data, "zscore"); ok && z > 0 {
		return math.Abs(z)
	}

	// Funding rate.
	if rate, ok := getFloat(data, "funding_rate"); ok {
		absRate := math.Abs(rate)
		switch {
		case absRate >= 0.002:
			return 6.0
		case absRate >= 0.001:
			return 4.0
		case absRate >= 0.0005:
			return 2.5
		}
	}

	// Liquidation total in millions.
	if millions, ok := getFloat(data, "total_liquidation_millions"); ok {
		switch {
		case millions >= 20:
			return 6.0
		case millions >= 10:
			return 4.0
		case millions >= 5:
			return 3.0
		}
	}

	// Open interest change % (port from Python: also reads change_pct).
	if oi, ok := getFloat(data, "change_pct"); ok {
		absOI := math.Abs(oi)
		switch {
		case absOI >= 30:
			return 5.0
		case absOI >= 15:
			return 3.0
		}
	}

	// Velocity change % (port from Python: same field, different thresholds).
	if vel, ok := getFloat(data, "change_pct"); ok {
		absVel := math.Abs(vel)
		switch {
		case absVel >= 3:
			return 4.0
		case absVel >= 1.5:
			return 2.5
		}
	}

	// Volume spike ratio.
	if ratio, ok := getFloat(data, "ratio"); ok {
		switch {
		case ratio >= 15:
			return 5.0
		case ratio >= 10:
			return 3.0
		}
	}

	// L/S ratio deviation from 1.
	if lsRatio, ok := getFloat(data, "ls_ratio"); ok && math.Abs(lsRatio-1) > 3 {
		return 5.0
	}

	// Long rate.
	if longRate, ok := getFloat(data, "long_rate"); ok {
		switch {
		case longRate >= 85:
			return 4.0
		case longRate >= 80:
			return 2.5
		}
	}

	return 0.0
}

// directionBias determines bullish (1) / bearish (-1) / neutral (0) per event type.
func directionBias(eventType string, data map[string]any) int {
	switch eventType {
	case "price_velocity":
		if pct, ok := getFloat(data, "change_pct"); ok {
			if pct > 0 {
				return 1
			}
			return -1
		}
		return 0
	case "volume_spike":
		return 0
	case "open_interest_change":
		return 0
	case "funding_rate_anomaly":
		if rate, ok := getFloat(data, "funding_rate"); ok {
			if rate > 0 {
				return -1 // positive funding = bearish
			}
			return 1 // negative funding = bullish
		}
		return 0
	case "long_short_ratio":
		if longRate, ok := getFloat(data, "long_rate"); ok {
			switch {
			case longRate > 60:
				return -1
			case longRate < 40:
				return 1
			default:
				return 0
			}
		}
		return 0
	case "liquidation":
		if longPct, ok := getFloat(data, "long_liquidation_pct"); ok {
			switch {
			case longPct > 60:
				return -1
			case longPct < 40:
				return 1
			default:
				return 0
			}
		}
		return 0
	}
	return 0
}

// ── Helpers ─────────────────────────────────────────────────────────────

// getFloat extracts a float64 value from a map.
func getFloat(data map[string]any, key string) (float64, bool) {
	v, ok := data[key]
	if !ok || v == nil {
		return 0, false
	}
	switch n := v.(type) {
	case float64:
		return n, true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case uint64:
		return float64(n), true
	case uint:
		return float64(n), true
	case int32:
		return float64(n), true
	case uint32:
		return float64(n), true
	case float32:
		return float64(n), true
	case uint8:
		return float64(n), true
	case int8:
		return float64(n), true
	case bool:
		if n {
			return 1, true
		}
		return 0, true
	case string:
		return 0, false // Python try/except would return 0, but we return not-ok
	default:
		return 0, false
	}
}

func allPositive(vals []int) bool {
	for _, v := range vals {
		if v <= 0 {
			return false
		}
	}
	return true
}

func allNegative(vals []int) bool {
	for _, v := range vals {
		if v >= 0 {
			return false
		}
	}
	return true
}

func sevScore(s string) int {
	switch s {
	case "LOW":
		return 1
	case "MEDIUM":
		return 2
	case "HIGH":
		return 3
	default:
		return 1
	}
}
